### ISSUE TITLE
Can’t copy-paste in Mac-companion-app

### ISSUE THREAD

**Issue by srmeuwese** [home-assistant/iOS#2154]
Hello there,

I saw that historically it wasn’t possible to copy-paste yaml in de Apple companion-app but seeing as were now two years later. Surprised to see it’s still not possible. 
would love to edit my automations within the app.

**Comment by johngianni** [comment]
I've tried all sort of things, including compiling the code to debug (but it's impossible with a free dev account)... In my opinion, it is not an easy fix, as I see a lot of other problems with scrolling, selecting, highlighting in the various pages. There's something wrong in the framework, I guess.

**Comment by ekeih** [comment]
Could be a duplicate of https://github.com/home-assistant/iOS/issues/1154

**Comment by Steve-gnome** [comment]
yup - major pain the a....

**Comment by MartinJohannesNilsen** [comment]
Still a problem

**Comment by dunxd** [comment]
Frustrating. Lots of scroll bars don't work either.  The File Editor is unusable without scrolling and copy/paste.

**Comment by toddbreyman** [comment]
How is this still a problem?

**Comment by dunxd** [comment]
> How is this still a problem?
See [
](https://github.com/home-assistant/iOS/issues/1154#issuecomment-1409734535)

**Comment by kcoffau** [comment]
issue persists in HAOS 2024.3 / Companion 2024.2 (2024.561)

Also copy paste from File Editor.

**Comment by ThermScissorpunch** [comment]
I also experience this.

**Comment by ZeFish** [comment]
Here also

**Comment by Rvervuurt** [comment]
How is this still a problem?! It's such basic functionality.

**Comment by bgoncal** [comment]
While Apple does not fix this WKWebView issue on Mac Catalyst apps, here is an workaround:
https://github.com/home-assistant/iOS/pull/3975

You will need to use shift + command + c/v , but better than nothing.

Also added buttons to top bar:

<img width="1982" height="1672" alt="Image" src="https://github.com/user-attachments/assets/1406f47b-e37a-400f-ae0a-de3853ca1f39" />

**Comment by dreyks** [comment]
still not working for me :( the Paste button works, but the copy one doesn't. (same with the shift+cmd+c/v) do I need to tweak any permission settings for the copy to work as well?

**Comment by bgoncal** [comment]
No, unfortunately if the hack doesn't work then we are back on no other workarounds available :/ I'll see if I can find more, but give it a try more times to find out if it never works or just on specific situation, for me it is now working very reliably

**Comment by dreyks** [comment]
oh, i see, different places behave differently. here's what i've got:
- copying a UI text
  - works every time directly for Cmd+C (as well as for the new copy button)
- copying input values inside the forms (like entity form)
  - same as the above, everything works
- copying text from Code editor (like in automations or dashboards)
  - Cmd+C **doesn't** work; trying to paste elsewhere after copying produces a system "beep" sound as if there's nothing in the clipboard, or it's in a wrong format
  - Shift+Cmd+C works
  - so does the button
- copying text from File editor
  - none of the Cmd+C/Shift+Cmd+C/button work, each of them producing the same state as with the Code editor (empty clipboard)

**Comment by kdhoe85** [comment]
Seems to be fixed in macOS 26.4 beta 2. I can now copy-paste without issues from and to the HA companion app. 
Can someone also test & confirm this?

**Comment by zang74** [comment]
As it seems entirely to revolve around copy (paste seems to act normal) here are two workarounds, one for the Mouse and one for Keyboard Shortcuts. This works within dashboard editing.

Mouse (free) : 
Don't use Right-click -> Copy. 

Use Right-click -> Share -> Copy. 

I do not know why this works from the share menu but it doesn't from the regular contextual menu, but at least it *does* work, even if it's an extra step.

Keyboard ($):
I use Keyboard Maestro and set up a simple Hot Key swap:

<img width="794" height="349" alt="Image" src="https://github.com/user-attachments/assets/8c1618ec-4e2e-409c-9e45-84d5071e8c9d" />

**Comment by zang74** [comment]
> Seems to be fixed in macOS 26.4 beta 2. I can now copy-paste without issues from and to the HA companion app. Can someone also test & confirm this?

Tested it, but it still does not work.

**Comment by maniakum** [comment]
It still doesn't work. It's absolutely frustrating when editing code, when you can't copy and paste and have to open the browser - horrible.

**Comment by Tartou69** [comment]
Hey. These New way ShiftCommandC or the buttons uper the window works but are limited to between 70 and 90 lines code.

How to fully copy a code ?
